src/audio/source_api_helper.c

```diff
@@ -0,0 +1,135 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2023 Intel Corporation. All rights reserved.
+//
+
+#include <sof/audio/source_api.h>
+#include <sof/audio/source_api_internal.h>
+
+inline void source_init(struct source *src, const struct source_ops *ops,
+			struct sof_audio_stream_params *audio_stream_params)
+{
+	src->ops = ops;
+	src->locked_read_frag_size = 0;
+	src->audio_stream_params = audio_stream_params;
+}
+
+inline size_t source_get_data_available(audio_src_hnd source)
+{
+	return source->ops->get_data_available(source);
+}
+
+inline int source_get_data(
+		audio_src_hnd source, size_t req_size,
+		void **data_ptr, void **buffer_start, void **buffer_end, size_t *buffer_size)
+{
+	int ret;
+
+	if (source->locked_read_frag_size)
+		return -EBUSY;
+
+	ret = source->ops->get_data(source, req_size, data_ptr,
+						 buffer_start, buffer_end, buffer_size);
+
+	if (!ret)
+		source->locked_read_frag_size = req_size;
+	return ret;
+}
+
+inline int source_put_data(audio_src_hnd source, size_t free_size)
+{
+	int ret;
+
+	/* Check if anything was obtained before for reading by source_get_data */
+	if (!source->locked_read_frag_size)
+		return -ENODATA;
+
+	/* limit size of data to be freed to previously obtained size */
+	if (free_size > source->locked_read_frag_size)
+		free_size = source->locked_read_frag_size;
+
+	ret = source->ops->put_data(source, free_size);
+
+	if (!ret)
+		source->locked_read_frag_size = 0;
+
+	source->num_of_bytes_processed += free_size;
+	return ret;
+}
+
+inline uint32_t source_get_num_of_processed_bytes(audio_src_hnd source)
+{
+	return source->num_of_bytes_processed;
+}
+
+inline void source_reset_num_of_processed_bytes(audio_src_hnd source)
+{
+	source->num_of_bytes_processed = 0;
+}
+
+inline enum sof_ipc_frame source_get_frm_fmt(audio_src_hnd source)
+{
+	return source->audio_stream_params->frame_fmt;
+}
+
+inline enum sof_ipc_frame source_get_valid_fmt(audio_src_hnd source)
+{
+	return source->audio_stream_params->valid_sample_fmt;
+}
+
+inline uint32_t source_get_rate(audio_src_hnd source)
+{
+	return source->audio_stream_params->rate;
+}
+
+inline uint32_t source_get_channels(audio_src_hnd source)
+{
+	return source->audio_stream_params->channels;
+}
+
+inline bool source_get_underrun(audio_src_hnd source)
+{
+	return source->audio_stream_params->underrun_permitted;
+}
+
+inline void source_set_valid_fmt(audio_src_hnd source,
+				      enum sof_ipc_frame valid_sample_fmt)
+{
+	source->audio_stream_params->valid_sample_fmt = valid_sample_fmt;
+	if (source->ops->on_audio_format_set)
+		source->ops->on_audio_format_set(source);
```

Comment:
Good point, 
To avoid creating a function pointer for every parameter I thought about change notification only. But yes, some validation should be there. Will think about it